metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala

```diff
@@ -44,47 +112,374 @@ class PackageProvider(private val buildTargets: BuildTargets) {
+      importer: Importer,
+      importParts: Vector[String],
+      oldPkgParts: Vector[String],
+      refsNames: Set[String],
+      newPkgName: String,
+  ) =
+    if (importParts == oldPkgParts) {
+      val importees = importer.importees
+      def isReferenced(importee: Importee) = importee match {
+        case Importee.Name(name) => refsNames.contains(name.value)
+        case Importee.Rename(from, _) => refsNames.contains(from.value)
+        case Importee.Unimport(name) => refsNames.contains(name.value)
+        case _ => false
+      }
+      val (refsImportees, refsIndices) = importees.zipWithIndex.filter {
+        case (i, _) =>
+          isReferenced(i)
+      }.unzip
+      if (refsImportees.length == 0) {
+        List.empty
+      } else {
+        val importeeStr = mkImportees(refsImportees.collect {
+          case Importee.Name(name) => name.value
+          case Importee.Rename(from, to) =>
+            s"${from.value} => ${to.value}"
+          case Importee.Unimport(name) => name.value
+        })
+        val newImportStr = {
+          val bracedImporteeStr = refsImportees match {
+            case List(Importee.Rename(_, _)) => s"{$importeeStr}"
+            case _ => importeeStr
           }
+          s"import $newPkgName.$bracedImporteeStr"
+        }
+        if (refsImportees.length == importer.importees.length) {
+          List(new TextEdit(importer.pos.toLsp, newImportStr))
+        } else {
+          // In this branch some importees should be removed from the import, but not all.
+          // Trying to remove them with commas
+          val indSet = refsIndices.toSet
+          val oldImportEdits =
+            importees.zip(importees.drop(1)).zipWithIndex.flatMap {
+              case ((i1, i2), ind) =>
+                val pos =
+                  if (indSet.contains(ind))
+                    Some(
+                      i1.pos.toLsp.copy(
+                        endLine = i2.pos.startLine,
+                        endCharacter = i2.pos.startColumn,
+                      )
+                    )
+                  else if (
+                    indSet.contains(ind + 1) && importees.length == ind + 2
+                  ) // cases like "import foo.{A, B, C}" when C is moved
+                    Some(
+                      i2.pos.toLsp.copy(
+                        startLine = i1.pos.endLine,
+                        startCharacter = i1.pos.endColumn,
+                      )
+                    )
+                  else None
+                pos.map(new TextEdit(_, ""))
+            }
+          val newImportPos = posNextToEndOf(importer.pos)
+          val newImportEdit = new TextEdit(newImportPos, newImportStr)
+          newImportEdit :: oldImportEdits
         }
-    } else {
-      None
+      }
+    } else List.empty
+
+  private def calcNewImportsForFileMove(
+      optTree: Option[Tree],
+      refsNames: Set[String],
+      newPkgName: String,
+  ) = {
```

Comment:
Can you add explicit types even for private functions? It helps reading code on GitHub, etc.

Reply:
Yep, thanks for pointing out